Ignore launchpad deposit data json file

vishaljhala opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/ConsenSys/teku/blob/master/CONTRIBUTING.md -->

## PR Description
Users commonly copy all the files generated by the deposit-cli into their teku keys directory, including the deposit_data json file. Since it's a .json file, Teku tries to load it as a keystore and fails to find a password file matching it resulting in confusing errors.

We should just ignore the deposit_data json file.

Sample content of deposit_data-1620858087.json:

We probably should check the content is at least vaguely like a deposit data file or at least not a keystore file before ignoring the file rather than just ignoring all deposit_data-*.json files.

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->
https://github.com/ConsenSys/teku/issues/3971

## Documentation

- [ ] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

ajsutton commented on validator/api/src/main/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocator.java:
```diff
@@ -80,12 +82,29 @@ private void parseEntry(final String keyFileName, final String passwordFileName)
     }
     if (keyFile.isFile()) {
-      pathMap.putIfAbsent(keyFile.toPath(), passwordFile.toPath());
+      if (!isDepositDataFile(keyFile)) pathMap.putIfAbsent(keyFile.toPath(), passwordFile.toPath());
```
nit: We always use braces on if statements:
```suggestion
      if (!isDepositDataFile(keyFile)) {
          pathMap.putIfAbsent(keyFile.toPath(), passwordFile.toPath());
      }
```

ajsutton commented on validator/api/src/main/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocator.java:
```diff
@@ -80,12 +82,29 @@ private void parseEntry(final String keyFileName, final String passwordFileName)
     }
   }
 
+  private boolean isDepositDataFile(final File keyFile) {
```
I'd probably filter by filename before checking the file contents. Deposit data files will always start with `deposit_data-`.

ajsutton commented on validator/api/src/main/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocator.java:
```diff
@@ -80,12 +82,29 @@ private void parseEntry(final String keyFileName, final String passwordFileName)
+      if (!listData.isEmpty() && listData.size() > 0) {
+        if (listData.get(0).containsKey("deposit_message_root")
+            && listData.get(0).containsKey("deposit_data_root")) return true;
+      }
```
Can probably simplify this to just a single `return` statement and not need the `if` statements (the `&&` operator short circuits so won't execute the right hand side if the left hand side returns false).

ajsutton commented on validator/api/src/main/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocator.java:
```diff
@@ -80,12 +82,29 @@ private void parseEntry(final String keyFileName, final String passwordFileName)
+            && listData.get(0).containsKey("deposit_data_root")) return true;
+      }
+    } catch (final Exception e) {
+      return false;
```
Would be worth logging a message at debug level in this case so there's some record that we failed to check the file.

ajsutton commented on validator/api/src/main/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocator.java:
```diff
@@ -111,7 +130,8 @@ void parseDirectory(final File keyDirectory, final File passwordDirectory) {
                 }
-                pathMap.putIfAbsent(path, passwordFile.toPath());
+                if (!isDepositDataFile(path.toFile()))
+                  pathMap.putIfAbsent(path, passwordFile.toPath());
```
nit: Braces. :)

ajsutton commented on validator/api/src/test/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocatorTest.java:
```diff
@@ -103,6 +104,27 @@ public void shouldIgnoreSomeFiles(@TempDir final Path tempDir) throws IOExceptio
+        tempDir,
+        Path.of("key", "deposit-data.json"),
+        Path.of("key", "a.json"),
+        Path.of("pass", "deposit-data.txt"),
```
The test should pass without this `deposit-data.txt` - there won't normally be a password file for the deposit data file.

ajsutton commented on validator/api/src/test/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocatorTest.java:
```diff
@@ -103,6 +104,27 @@ public void shouldIgnoreSomeFiles(@TempDir final Path tempDir) throws IOExceptio
+    createFolders(tempDir, "key", "pass");
+    createFiles(
+        tempDir,
+        Path.of("key", "deposit-data.json"),
```
Probably worth making this a bit more realistic and name the file `deposit_data-1620858087.json` since that's the type of name the deposit cli actually generates.

ajsutton commented on validator/api/src/test/java/tech/pegasys/teku/validator/api/KeyStoreFilesLocatorTest.java:
```diff
@@ -228,6 +250,29 @@ private void createFiles(final Path tempDir, Path... paths) throws IOException {
+
+    byte[] data = jsonData.getBytes(UTF_8);
+    File file = depositDataFilePath.toFile();
+    if (file != null) {
```
I'd omit this null check - if the file is null for some reason we want the test to fail anyway.

ajsutton reviewed (commented):
Thanks for submitting this, it's looking good. I've left a bunch of comments but they're nearly all just little details - the basic approach looks right to me.

ajsutton reviewed (approved):
LGTM. Thanks for this. Just need to work out what's going on with the CLA so we can merge it in.

vishaljhala committed:
Ignore launchpad deposit data json files

ajsutton committed:
Merge branch 'master' into 3971_ignore_launchpad_deposit_data_json_file

ajsutton committed:
Update changelog.
Add debug message when deposit data file parsing fails.
Minor code simplification.
